### PR TITLE
First variant of HW04

### PR DESCRIPTION
**Pull request by A-nastasi-a**

### CODE REVIEW

File: HW04/anastasia_evsihneyeva/lib/homework.rb [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+    @description = data[:desription]
+    @student = data[:student]
+    @mentor = mentor
+    @status = 'new'
+    @answers = []
+    @to_check = []
+  end
+
+  def add_answer!(answer)
+    @answers << answer
```
Comment (B0RlS) [inline HW04/anastasia_evsihneyeva/lib/homework.rb:15]: add whatever you are using to attr_reader
```suggestion
    answers << answer
```

File: HW04/anastasia_evsihneyeva/lib/mentor.rb [hunk 2]
```diff
@@ -0,0 +1,38 @@
...
+
+class Mentor
+  def initialize(data)
+    @name = data[:name]
+    @surname = data[:surname]
+    @students = []
+  end
+
+  def add_homework(data)
+    Notification.new(log: "-New homework was added! Name #{data[:title]}")
```
Comment (B0RlS) [inline HW04/anastasia_evsihneyeva/lib/mentor.rb:11]: Perhaps we should first create a homework before creating a notification.

File: HW04/anastasia_evsihneyeva/lib/mentor.rb [hunk 3]
```diff
@@ -0,0 +1,38 @@
...
+    name = "#{@name} #{@surname}"
+    Homework.new(data, name)
```
Comment (B0RlS) [inline HW04/anastasia_evsihneyeva/lib/mentor.rb:13]: It seems to me that homework should contain information about the mentor who created it and possibly about the student to whom this homework was given.

File: HW04/anastasia_evsihneyeva/lib/mentor.rb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+    Notification.new(log: "-Mentor #{@name} #{@surname} subscribed to #{student.name}")
+    @students << student
+  end
+
+  def notifications
+    Notification.new
+  end
+
+  def mark_as_read!
+    Notification.new(log: '-Mark_as_read!')
```
Comment (B0RlS) [inline HW04/anastasia_evsihneyeva/lib/mentor.rb:26]: If you are already doing such functionality, then it is expected that you will have a functionality that will store unread notifications, and then clear them after reading

File: HW04/anastasia_evsihneyeva/lib/notification.rb [hunk 5]
```diff
@@ -0,0 +1,12 @@
+# frozen_string_literal: true
+
+require 'singleton'
+
+class Notification
+  attr_reader :log
+
+  def initialize(log = { log: nil })
+    @log = log
+    puts @log[:log]
```
Comment (B0RlS) [inline HW04/anastasia_evsihneyeva/lib/notification.rb:10]:
```suggestion
    puts @log[:log]
```
it's a bad idea to use it here

File: HW04/anastasia_evsihneyeva/lib/student.rb [hunk 6]
```diff
@@ -0,0 +1,42 @@
...
+
+  def initialize(data)
+    @name = data[:name]
+    @surname = data[:surname]
+    @homeworks = []
+  end
+
+  def name
+    "#{@name} #{@surmane}"
+  end
```
Comment (B0RlS) [inline HW04/anastasia_evsihneyeva/lib/student.rb:14]:
```suggestion
```
use attr_reader

File: HW04/anastasia_evsihneyeva/lib/student.rb [hunk 7]
```diff
@@ -0,0 +1,42 @@
...
+
+  def notifications
```
Comment (B0RlS) [inline HW04/anastasia_evsihneyeva/lib/student.rb:16]: I do not think that this functionality should be used in the student

File: HW04/anastasia_evsihneyeva/lib/student.rb [hunk 8]
```diff
@@ -0,0 +1,42 @@
...
+  end
+
+  def to_work!(homework)
+    Notification.new(log: "-Student #{@name} #{@surname} working on #{homework.title}")
+    @homeworks << homework
+  end
+
+  def add_answer!(homework, answer)
+    text = "#{@name} #{@surname}: #{answer}"
+    homework.add_answer!(text)
```
Comment (B0RlS) [inline HW04/anastasia_evsihneyeva/lib/student.rb:35]:
```suggestion
    homework.add_answer!(answer)
```

File: HW04/anastasia_evsihneyeva/lib/homework.rb (hunk 1 above)
Comment (A-nastasi-a) [reply HW04/anastasia_evsihneyeva/lib/homework.rb:15]: But how can I do list attr_reader? I can do it only with symbols and if I try to use << with symbol appears an error...

File: HW04/anastasia_evsihneyeva/lib/notification.rb [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+  attr_accessor :logs
+
+  def initialize(text = nil)
+    @logs =  []
+    @logs << text if text
+  end
+
+  def mark_as_read!
+    @logs.clear
+  end
```
Comment (B0RlS) [inline HW04/anastasia_evsihneyeva/lib/notification.rb:13]: This is not the responsibility of this class (transfer this to the mentor and student class)

File: HW04/anastasia_evsihneyeva/lib/mentor.rb [hunk 10]
```diff
@@ -0,0 +1,48 @@
...
+    notification_list << notification
+  end
+
+  def mark_as_read!
+    notification_list.clear
+  end
+
+  def add_homework(title:, description:, student:)
+    homework = Homework.new(title, description, student, self)
+    homework.student.notify(Notification.new("Homework #{title} was added to #{student.name} by mentor #{name}"))
```
Comment (B0RlS) [inline HW04/anastasia_evsihneyeva/lib/mentor.rb:29]: It seems to me that the student should not notify himself when the mentor adds homework, make this method on the mentor and call it to add notification to the student

File: HW04/anastasia_evsihneyeva/lib/student.rb [hunk 11]
```diff
@@ -0,0 +1,46 @@
...
+
+  def mark_as_read!
+    notification_list.clear
+  end
+
+  def homeworks
+    all_hw.each do |hw|
+      puts "#{hw.title} - #{hw.description}" if hw.status != 'Finished'
+    end
+  end
```
Comment (B0RlS) [inline HW04/anastasia_evsihneyeva/lib/student.rb:30]:
```suggestion
```
You have all_hw and you can return it and see the data, you don't need to add output of some data from puts to the business logic, you can do it in app.rb if necessary

**Review by vladhilko — APPROVED** [review]

**Pull request merged**